File: c/rna-transcription/rna_transcription.c
#include "rna_transcription.h"

char *to_rna(const char *dna) {
    size_t dna_length = strlen(dna);
    char *rna = calloc(dna_length + 1, sizeof(char));

    if (dna_length == 0) {
        return rna;
    }

    for (size_t i = 0; i < dna_length; i++) {
        switch (toupper(dna[i])) {
        case 'G':
            rna[i] = 'C';
            break;
        case 'C':
            rna[i] = 'G';
            break;
        case 'T':
            rna[i] = 'A';
            break;
        case 'A':
            rna[i] = 'U';
            break;
        }

        rna[dna_length] = '\0';
    }

    return rna;
}

File: dart/rna-transcription/lib/rna_transcription.dart
// Transcribes a DNA strand to an RNA strand.
// https://exercism.org/tracks/dart/exercises/anagram

typedef nucleotide_t = String;
typedef nucleotides_t = List<nucleotide_t>;
typedef dna_strand_t = String;
typedef rna_strand_t = String;
typedef dna_to_rna_t = Map<nucleotide_t, nucleotide_t>;
typedef rna_to_dna_t = Map<nucleotide_t, nucleotide_t>;

class RnaTranscription {
  dna_to_rna_t dna_to_rna = {
    'G': 'C',
    'C': 'G',
    'T': 'A',
    'A': 'U',
  };

  rna_strand_t toRna(dna_strand_t dna_strand) {
    rna_strand_t rna_strand = "";

    nucleotides_t dna_nucleotides = dna_strand.split('');

    var iter = dna_nucleotides.iterator;

    while (iter.moveNext()) {
      nucleotide_t dna_nucleotide = iter.toString();
      rna_strand += dna_to_rna[dna_nucleotide].toString();
    }

    return rna_strand;
  }
}
